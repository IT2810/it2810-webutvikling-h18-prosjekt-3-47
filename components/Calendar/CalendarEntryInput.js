import React from 'react';
import {StyleSheet, Text, View, TouchableOpacity, TextInput, Button, DatePickerAndroid, Platform, AsyncStorage} from 'react-native';
import Modal from "react-native-modal";


export default class CalendarEntryInput extends React.Component {

    constructor (props){
        super(props);

        this.state = {
            isModalVisible: false,
            newEventDate: '2018-10-01',
            newEventText: null
        };

        this.requestClose = this.requestClose.bind(this);
        this.sendData = this.sendData.bind(this);
    }


    _toggleModal = () =>
        this.setState({ isModalVisible: !this.state.isModalVisible });
    
    requestClose() {
        this.setState({ isModalVisible: false })
    }

    sendData() {
        console.log('newEventText:', this.state.newEventText);
        console.log('newEventDate:', this.state.newEventDate);

        if (this.state.newEventDate && this.state.newEventText) {
            this.props.callback({
                newEventDate: this.state.newEventDate,
                newEventText: this.state.newEventText,
            })
        }
    }

    render() {

        return (
            <View style={styles.view}>
                <TouchableOpacity onPress={this._toggleModal}>
                    <Text>Show Modal</Text>
                </TouchableOpacity>
                <Modal
                    style={styles.modalContainer}
                    isVisible={this.state.isModalVisible}
                    onBackdropPress={this.requestClose}
                    onSwipe={this.requestClose}
                    swipeDirection="left"
                    onRequestClose={this.requestClose}
                    >

                    <View style={styles.modalView}>
                        <Text>Hello!</Text>

                        <TextInput
                            style={{height: 40, borderColor: 'gray', borderWidth: 1}}
                            onChangeText={(text) => this.setState({newEventText: text})}
                        />

                        <TouchableOpacity onPress={this._toggleModal}>
                            <Text>Hide me!</Text>
                        </TouchableOpacity>

                        <Button style={styles.button}
                                onPress={this.sendData}
                                title="Submit!"
                                color="#841584"
                                accessibilityLabel="Submit the new entry"
                        />
                    </View>
                </Modal>
            </View>
        )
    }
}

const styles = StyleSheet.create({
    view: {
        flex: 1,
        alignItems:'center'
    },
    modalContainer: {
      flex: 1,
      alignItems: 'center'
    },
    modalView: {
        width: '80%',
        flex: 1,
        backgroundColor: '#fff'
    },
    button: {

    }
});
