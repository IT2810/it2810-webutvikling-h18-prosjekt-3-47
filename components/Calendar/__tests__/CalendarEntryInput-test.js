import 'react-native';
import React from 'react';
import CalendarEntryInput from '../CalendarEntryInput';

import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';

jest.mock('react-native-modal-datetime-picker');
// jest.mock('react-native-material-textfield');

const momentMock = {
    format(){return dateString;}
};

const dateString = '2018-02-07';

describe('Rendering', () => {
    it('renders correctly with renderer', () => {
        console.log(momentMock);
        const calendarEntryInput = renderer.create(
            <CalendarEntryInput
                defaultDate={momentMock}
            />);

        expect(calendarEntryInput).toMatchSnapshot();
    });


    it('renders correctly with ShallowRenderer', () => {
        const shallowRenderer = new ShallowRenderer();
        const calendarEntryInput = shallowRenderer.render(<CalendarEntryInput
            defaultDate={momentMock}
        />);

        expect(calendarEntryInput).toMatchSnapshot();
    });
});


describe('Logical', () => {
    it('should have empty state and date prop', () => {

        let calendarEntryInput = renderer.create(
            <CalendarEntryInput callback={(data) => {console.log('callback!! ', data)}}
                                defaultDate={momentMock}
            />
        );
        let calendarEntryInputInstance = calendarEntryInput.getInstance();

        expect(calendarEntryInputInstance.state.newEventText).toEqual('');
        expect(calendarEntryInputInstance.state.newEventDate).toEqual(null);
        expect(calendarEntryInputInstance.state.isModalVisible).toEqual(false);
        expect(calendarEntryInputInstance.state.isDateTimePickerVisible).toEqual(false);
        expect(calendarEntryInputInstance.props.defaultDate.format()).toEqual(dateString);
    });

    it('should open and close Modal', () => {
        let calendarEntryInput = renderer.create(
            <CalendarEntryInput callback={(data) => {console.log('callback!! ', data)}}
                                defaultDate={momentMock}
            />
        );
        let calendarEntryInputInstance = calendarEntryInput.getInstance();

        calendarEntryInputInstance.requestShowModal();
        expect(calendarEntryInputInstance.state.isModalVisible).toEqual(true);

        calendarEntryInputInstance.requestCloseModal();
        expect(calendarEntryInputInstance.state.isModalVisible).toEqual(false);
    });
});


describe('input validation', () => {

    let calendarEntryInput = renderer.create(
        <CalendarEntryInput callback={(data) => {console.log('callback!! ', data)}}
                            defaultDate={momentMock}
        />
    );
    let calendarEntryInputInstance = calendarEntryInput.getInstance();

    it('should not allow pure whitespace', () => {

        expect(calendarEntryInputInstance.validateInput('                  ' +
            '' +
            '' +
            '\n    ', 'submit')
        ).toEqual(false);
    });

    it('should not allow empty string', () => {
        expect(calendarEntryInputInstance.validateInput('', 'submit')).toEqual(false);
    });

    it('should not allow strings with less than three characters', () => {
        expect(calendarEntryInputInstance.validateInput('      dd    ', 'submit')).toEqual(false);
    });

    it('should allow strings over length of two', () => {
        expect(calendarEntryInputInstance.validateInput('      ddd    ', 'submit')).toEqual(true);
    });

    // It only strips leading and trailing whitespace, so this one should be accepted:
    it('should allow strings over length of two', () => {
        expect(calendarEntryInputInstance.validateInput('      å e    ', 'submit')).toEqual(true);
    });
});


describe('callbacks and such', () => {
    it('should get a callback with the correct data', done => {
        const eventText = '     \n\n\n   Aaaassdææaø112-w33a##sdad sdd sd 323 sdsdø ++ +   \n    ';

        function callback(data) {
            expect(data.dateString).toEqual(dateString);
            expect(data.text).toEqual(eventText.trim());
            done();
        }

        let calendarEntryInput = renderer.create(
            <CalendarEntryInput callback={callback}
                                defaultDate={momentMock}
            />
        );

        let calendarEntryInputInstance = calendarEntryInput.getInstance();
        calendarEntryInputInstance.setState({
            newEventText: eventText,
            newEventDate: momentMock
        },() => {
            calendarEntryInputInstance.sendData();
        });
    });
});


