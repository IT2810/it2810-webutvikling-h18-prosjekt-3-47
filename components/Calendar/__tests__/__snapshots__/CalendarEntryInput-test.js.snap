// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering renders correctly with ShallowRenderer 1`] = `
<Component
  style={
    Array [
      Object {
        "alignItems": "center",
      },
      Object {
        "flex": 0,
      },
    ]
  }
>
  <TouchableOpacity
    accessibilityLabel="Legg til en ny kalenderhendelse"
    activeOpacity={0.2}
    onPress={[Function]}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 20,
        }
      }
    >
      Legg til ny hendelse
    </Text>
  </TouchableOpacity>
  <ReactNativeModal
    animationIn="slideInUp"
    animationInTiming={300}
    animationOut="slideOutDown"
    animationOutTiming={300}
    avoidKeyboard={false}
    backdropColor="black"
    backdropOpacity={0.7}
    backdropTransitionInTiming={300}
    backdropTransitionOutTiming={300}
    hideModalContentWhileAnimating={false}
    isVisible={false}
    onBackButtonPress={[Function]}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalShow={[Function]}
    onRequestClose={[Function]}
    onSwipe={[Function]}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
      }
    }
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeDirection="down"
    swipeThreshold={100}
    useNativeDriver={false}
  >
    <ScrollViewMock
      style={
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
          "padding": "2.5%",
          "width": "85%",
        }
      }
    >
      <TouchableOpacity
        accessibilityLabel="Skjul ny-kalenderhendelse-popupen"
        activeOpacity={0.2}
        onPress={[Function]}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
          Skjul popup
        </Text>
      </TouchableOpacity>
      <TextField
        accessibilityLabel="Skriv inn tekstbeskrivelse av ny kalenderhendelse"
        activeLineWidth={2}
        animationDuration={225}
        autoCapitalize="sentences"
        baseColor="rgba(0, 0, 0, .38)"
        disableFullscreenUI={true}
        disabled={false}
        disabledLineType="dotted"
        disabledLineWidth={1}
        editable={true}
        error={null}
        errorColor="rgb(213, 0, 0)"
        fontSize={18}
        inputContainerPadding={8}
        label="Hendelsesbeskrivelse"
        labelFontSize={16}
        labelHeight={32}
        labelPadding={4}
        lineWidth={0.5}
        multiline={true}
        onChangeText={[Function]}
        placeholder="Møte med Siv Jensen"
        textColor="rgba(0, 0, 0, .87)"
        tintColor="rgb(0, 145, 234)"
        titleFontSize={16}
        underlineColorAndroid="transparent"
        value=""
      />
      <Text
        accessibilityLabel="Valgt dato for den nye kalenderhendelsen er 2018-02-07"
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#505050",
            "fontSize": 16,
            "marginBottom": 10,
            "marginLeft": 5,
          }
        }
      >
        Valgt dato: 
        2018-02-07
      </Text>
      <TouchableOpacity
        accessibilityLabel="Åpne datovelgeren for å velge dato for den nye kalenderhendelsen"
        activeOpacity={0.2}
        onPress={[Function]}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
           Åpne datovelgeren
        </Text>
      </TouchableOpacity>
      <TouchableOpacity
        accessibilityLabel="Lagre kalenderhendelsen"
        activeOpacity={0.2}
        onPress={[Function]}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 25,
              "marginBottom": 15,
              "marginTop": 10,
            }
          }
        >
           Lagre 
        </Text>
      </TouchableOpacity>
    </ScrollViewMock>
  </ReactNativeModal>
  <CustomDatePickerIOS
    cancelTextIOS="Avbryt"
    confirmTextIOS="Velg"
    isVisible={false}
    locale="nb"
    mode="date"
    neverDisableConfirmIOS={false}
    onCancel={[Function]}
    onConfirm={[Function]}
    onDateChange={[MockFunction]}
    onHideAfterConfirm={[Function]}
    reactNativeModalPropsIOS={Object {}}
    titleIOS="Velg en dato"
  />
</Component>
`;

exports[`Rendering renders correctly with renderer 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
      },
      Object {
        "flex": 0,
      },
    ]
  }
>
  <View
    accessibilityLabel="Legg til en ny kalenderhendelse"
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 20,
        }
      }
    >
      Legg til ny hendelse
    </Text>
  </View>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onModalHide={[Function]}
    onRequestClose={[Function]}
    onSwipe={[Function]}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeDirection="down"
    swipeThreshold={100}
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
    />
    <View
      hideModalContentWhileAnimating={false}
      onModalHide={[Function]}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onRequestClose={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      onSwipe={[Function]}
      pointerEvents="box-none"
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "left": 0,
          "margin": 37.5,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeDirection="down"
      swipeThreshold={100}
    />
  </Modal>
</View>
`;
